lib/image-stream/parse-partitions.js

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2016 resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const Bluebird = require('bluebird');
+const MBR = require('mbr');
+const GPT = require('gpt');
+
+const MAX_BYTES = 8192;
+const INITIAL_LENGTH = 0;
+const INITIAL_BLOCK_SIZE = 512;
+const MAX_BLOCK_SIZE = 4096;
+
+/**
+ * Attempt to parse the GPT from various block sizes
+ * @private
+ * @param {Buffer} buffer - Buffer
+ * @returns {GPT|null}
+ */
+const parseGPT = (buffer) => {
+
+  let blockSize = INITIAL_BLOCK_SIZE;
+  let gpt = null;
+
+  // Attempt to parse the GPT from several offsets,
+  // as the block size of the image may vary (512,1024,2048,4096);
+  // For example, ISOs will usually have a block size of 4096,
+  // but raw images a block size of 512 bytes
+  while (blockSize <= MAX_BLOCK_SIZE) {
+    try {
+      gpt = GPT.parse(buffer.slice(blockSize));
+    } catch (error) {
+      // Ignore error
+    }
+    if (gpt) {
+      return gpt;
+    }
+    blockSize += blockSize;
+  }
+
+  return gpt;
+
+};
+
+/**
+ * Attempt to parse the MBR & GPT from a given buffer
+ * @private
+ * @param {Object} image - Image metadata
+ * @param {Buffer} buffer - Buffer
+ * @returns {undefined}
+ */
+const parsePartitions = (image, buffer) => {
+
+  let mbr = null;
+  let gpt = null;
+
+  try {
+    mbr = MBR.parse(buffer);
```

Comment:
I realise it goes somewhat against @jviotti 's earlier suggestion, but IMHO this should probably be `mbr = MBR.parse(buffer.slice(0, 512));` (with the numbers replaced with suitable constants). Perhaps you could even import `MBR_OFFSET` and `MBR_LENGTH` constants from your `mbr` module?